Add ApplyUnitary operation

fedimser opened the pull request:
This PR adds Q# operation which applies arbitrary gate specified by 2^n x 2^n unitary matrix on a qubit register, as requested in #291.

It's based on an algorithm which I have implemented in Python in [this repository](https://github.com/fedimser/quantum_decomp).

Implementation consists of 3 parts:
1. Implementing 1-qubit gate specified by 2x2 matrix - in native Q#.
2. Matrix decomposition algorithm, which decomposes unitary matrix as product of two-level unitary matrices with indices differing exactly in 1 bit - in C#.
3. Q# operation which uses decomposition in (2) and implements arbitrary n-qubit unitary via multiple calls to controlled (1) operation and X gates.

Complexity of this algorithm is O(N^3) and size of output is O(N^2), where N is size of matrix (N=2^n, where n is number of qubits).

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.cs:
```diff
@@ -0,0 +1,75 @@
+            }
+
+            private QArray<QArray<QArray<Complex>>> matricesToQs(List<Complex[,]> matrices) {
+                int n = (int)matrices.Count;
```
```suggestion
                int n = matrices.Count;
```

`Count` already returns `int`.

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.cs:
```diff
@@ -0,0 +1,75 @@
+
+            private QArray<QArray<QArray<Complex>>> matricesToQs(List<Complex[,]> matrices) {
+                int n = (int)matrices.Count;
+                QArray<QArray<Complex>>[] result = new QArray<QArray<Complex>>[n];
```
```suggestion
                var result = new QArray<QArray<Complex>>[n];
```

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.cs:
```diff
@@ -0,0 +1,75 @@
+
+            private QArray<QArray<Complex>> matrixToQs(Complex[,] b) {
+                long n = b.GetLength(0);
+                QArray<Complex>[] a = new QArray<Complex>[n];
```
```suggestion
                var a = new QArray<Complex>[n];
```

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.cs:
```diff
@@ -0,0 +1,75 @@
+            public override Func<IQArray<IQArray<Complex>>, (IQArray<IQArray<IQArray<Complex>>>, IQArray<IQArray<long>>)> __Body__ => TwoLevelDecomposeGray;
+        }
+    }
+}
```
```suggestion
}

```

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.qs:
```diff
@@ -0,0 +1,80 @@
+        let n = Length(matrices);
+        
+        mutable prevFlipMask = 0;
+        for (i in 0..n-1) {
```
You could write `for ((matrix, index) in TwoLevelDecomposition(unitary))` here and directly access `matrix` and `index` without a loop index variable.

fedimser replied:
Result of that function is not array of tuples, but tuple of arrays. However, maybe I should make it array of tuples.

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.qs:
```diff
@@ -0,0 +1,80 @@
+        for (i in 0..n-1) {
+            let matrix = matrices[i]; // Matrix of FC gate.
+            let index1 = idx[i][0];   // Indices of non-tivial 2x2 submatrix.
+            let index2 = idx[i][0];
```
Here, `index2 == index1`. Is that intended?

msoeken commented on Standard/tests/Synthesis/UnitaryDecompositionTest.qs:
```diff
@@ -0,0 +1,84 @@
+
+    // dotnet test tests --filter "Name~ApplyUnitary"
+    
+    operation ApplyUnitaryToRegister(matrix: Complex[][], qubits: Qubit[]) : Unit {
```
```suggestion
    internal operation ApplyUnitaryToRegister(matrix: Complex[][], qubits: Qubit[]) : Unit {
```

msoeken commented on Standard/tests/Synthesis/UnitaryDecompositionTest.qs:
```diff
@@ -0,0 +1,84 @@
+        ApplyUnitary(matrix, LittleEndian(qubits));
+    }
+
+    operation CheckOperation(matrix: Complex[][], expected: (Qubit[] => Unit is Adj)) : Unit {
```
```suggestion
    internal operation CheckOperation(matrix: Complex[][], expected: (Qubit[] => Unit is Adj)) : Unit {
```

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.qs:
```diff
@@ -0,0 +1,80 @@
+        if(AbsD(phi) > 1e-9) {R1(phi, qubit);}
+    }
+
+    function TwoLevelDecomposition(unitary: Complex[][]) : (Complex[][][], Int[][]) {
```
```suggestion
    internal function TwoLevelDecomposition(unitary: Complex[][]) : (Complex[][][], Int[][]) {
```

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.qs:
```diff
@@ -0,0 +1,80 @@
+        body intrinsic;
+    }
+
+    operation ApplyFlips(flipMask: Int, qubits : LittleEndian) : Unit is Adj + Ctl  {
```
```suggestion
    internal operation ApplyFlips(flipMask: Int, qubits : LittleEndian) : Unit is Adj + Ctl  {
```

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.cs:
```diff
@@ -0,0 +1,75 @@
+using Microsoft.Quantum.Math;
+
+namespace Microsoft.Quantum.Synthesis {
+    public partial class TwoLevelDecomposition {
```
```suggestion
    internal partial class TwoLevelDecomposition {
```

msoeken reviewed (commented):
Hi @fedimser, you're initial commits look interesting! I left some initial comments.

fedimser commented:
Thanks for the review.
This PR is work in progress, so I converted it to draft.

fedimser commented:
@msoeken, can you please advice which linear algebra library I can use in the C# code?
All I need is to be able to represent complex matrices and matrix multiplication. I can implement it from scratch on top of C# arrays, but I would prefer to use some C# equivalent of NumPy.

msoeken commented:
@fedimser I don't have much experience with NumPy alternatives in .NET, but you can see if [NumSharp](https://github.com/SciSharp/NumSharp) has the operations you need.

msoeken commented on Standard/src/Synthesis/MatrixUtils.cs:
```diff
@@ -0,0 +1,68 @@
+    internal class MatrixUtils
+    {
+        // Checks whether given matrix is unitary.
+        public static bool isMatrixUnitary(Complex[,] matrix)
```
```suggestion
        public static bool IsMatrixUnitary(Complex[,] matrix)
```

C# language style uses capital method names.

msoeken commented on Standard/src/Synthesis/MatrixUtils.cs:
```diff
@@ -0,0 +1,68 @@
+                        dotProduct += matrix[i, k] * Complex.Conjugate(matrix[j, k]);
+                    }
+                    Complex expectedDotProduct = (i == j) ? 1.0 : 0.0;
+                    if ((dotProduct - expectedDotProduct).Magnitude > 1e-9)
```
I suggest to extract the threshold as a parameter with a default value. Most of the accuracy thresholds in Q# are `1e-10`.

fedimser replied:
Done.

msoeken commented on Standard/src/Synthesis/MatrixUtils.cs:
```diff
@@ -0,0 +1,68 @@
+        }
+
+        // Converts matrix from C# array to Q# array.
+        public static QArray<QArray<Quantum.Math.Complex>> matrixToQs(Complex[,] b)
```
```suggestion
        public static QArray<QArray<Quantum.Math.Complex>> MatrixToQs(Complex[,] b)
```

msoeken commented on Standard/src/Synthesis/MatrixUtils.cs:
```diff
@@ -0,0 +1,68 @@
+        }
+
+        // Converts square matrix from Q# array to C#.
+        public static Complex[,] squareMatrixFromQs(IQArray<IQArray<Quantum.Math.Complex>> a)
```
```suggestion
        public static Complex[,] SquareMatrixFromQs(IQArray<IQArray<Quantum.Math.Complex>> a)
```

msoeken commented on Standard/src/Synthesis/TwoLevelUnitary.cs:
```diff
@@ -0,0 +1,71 @@
+
+        public TwoLevelUnitary(Complex[,] mx, int i1, int i2)
+        {
+            Debug.Assert(MatrixUtils.isMatrixUnitary(mx), "Matrix is not unitary");
```
```suggestion
            Debug.Assert(MatrixUtils.IsMatrixUnitary(mx), "Matrix is not unitary");
```

msoeken commented on Standard/src/Synthesis/TwoLevelUnitary.cs:
```diff
@@ -0,0 +1,71 @@
+
+        public bool isIdentity()
+        {
+            return (mx[0, 0] - 1).Magnitude < 1e-9 && mx[0, 1].Magnitude < 1e-9 &&
```
See comment above thresholds above.

fedimser replied:
Done.

msoeken commented on Standard/src/Synthesis/TwoLevelUnitary.cs:
```diff
@@ -0,0 +1,71 @@
+            return (MatrixUtils.matrixToQs(this.mx), this.i1, this.i2);
+        }
+    }
+}
```
```suggestion
}

```

msoeken commented on Standard/src/Synthesis/TwoLevelUnitary.cs:
```diff
@@ -0,0 +1,71 @@
+        // Converts to tuple to be passed to Q#.
+        public (IQArray<IQArray<Quantum.Math.Complex>>, long, long) toQsharp()
+        {
+            return (MatrixUtils.matrixToQs(this.mx), this.i1, this.i2);
```
```suggestion
            return (MatrixUtils.MatrixToQs(this.mx), this.i1, this.i2);
```

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.cs:
```diff
@@ -0,0 +1,149 @@
+
+            // Returns permutation of numbers from 0 to n-1 such as any two consequent number 
+            // differ in exactly one bit.
+            private static int[] grayCode(int n)
```
Great implementation!

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.qs:
```diff
@@ -0,0 +1,88 @@
+        // If this is not special unitary, apply R1 to correct global phase.
+        let det = MinusC(TimesC(u[0][0], u[1][1]), TimesC(u[0][1], u[1][0]));
+        let phi = ArgComplex(det);
+        if(AbsD(phi) > 1e-9) {R1(phi, qubit);}
```
```suggestion
        if (AbsD(phi) > 1e-9) { R1(phi, qubit); }
```

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.qs:
```diff
@@ -0,0 +1,88 @@
+    internal function GetFlipMasks(decomposition: (Complex[][], Int, Int)[], 
+                                   allQubitsMask: Int) : Int[] {
+        let n = Length(decomposition);
+        mutable flipMasks = new Int[n+1];
+        set flipMasks w/= 0 <- 0;
```
```suggestion
        mutable flipMasks = ConstantArray(n + 1, 0);
```

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.qs:
```diff
@@ -0,0 +1,88 @@
+        let n = Length(decomposition);
+        mutable flipMasks = new Int[n+1];
+        set flipMasks w/= 0 <- 0;
+        for (i in 0..n-1) {
+            let (matrix, i1, i2) = decomposition[i];
+            set flipMasks w/= (i+1) <- (allQubitsMask - i2); 
+        }
```
```suggestion
        for ((i, (_, _, i2)) in Enumerated(decomposition)) {
            set flipMasks w/= (i+1) <- (allQubitsMask - i2); 
        }
```

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.qs:
```diff
@@ -0,0 +1,88 @@
+
+    // Applies X gates to qubits speicifed by flipMask.
+    internal operation ApplyFlips(flipMask: Int, qubits : LittleEndian) : Unit is Adj + Ctl  {
+        for (qubitId in 0..Length(qubits!)-1) {
+            if ((flipMask &&& (1 <<< qubitId)) != 0) { X(qubits![qubitId]); }    
+        }
```
```suggestion
        ApplyXorInPlace(flipMask, qubits);
```
I think that these two do the very same thing.

fedimser replied:
Nice function! In that case I don't ned ApplyFlips and can call ApplyXorInPlace instead.

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.qs:
```diff
@@ -0,0 +1,88 @@
+        let decomposition = TwoLevelDecomposition(unitary);
+        let flipMasks = GetFlipMasks(decomposition, allQubitsMask);
+        
+        for (i in 0..Length(decomposition)-1) {
+            // i1, i2 - indices of non-trivial 2x2 submatrix of two-level unitary matrix being 
+            // applied. 
+            // i1 and i2 differ in exactly one bit; i1 < i2.
+            // matrix - 2x2 non-trivial unitary submatrix of said two-level unitary.
+            let (matrix, i1, i2) = decomposition[i];
```
```suggestion
        // i1, i2 - indices of non-trivial 2x2 submatrix of two-level unitary matrix being 
        // applied. 
        // i1 and i2 differ in exactly one bit; i1 < i2.
        // matrix - 2x2 non-trivial unitary submatrix of said two-level unitary.
        for ((i, (matrix, i1, i2)) in Enumerated(decomposition)) {
```

msoeken reviewed (changes requested):
Thanks a lot @fedimser, this is great work! I left some comments, mostly formatting and some suggestions.

cgranade commented on Standard/src/Synthesis/UnitaryDecomposition.qs:
```diff
@@ -0,0 +1,88 @@
+        body intrinsic;
+    }
+
+    // For every 2-level unitary calculates "flip mask", which denotes qubit which should 
+    // be inverted before and after applying corresponding 1-qubit gate.
+    // For convenience prepends result with 0.
```
It may be helpful to turn this into a brief API documentation comment.

```suggestion
    /// # Summary
    /// For every 2-level unitary calculates "flip mask", which denotes qubit which should 
    /// be inverted before and after applying corresponding 1-qubit gate.
    /// For convenience prepends result with 0.
```

cgranade commented on Standard/tests/Synthesis/UnitaryDecompositionTest.qs:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    @Test("QuantumSimulator")
+    operation ApplyUnitary_Identity () : Unit {
```
In general, the Q# style guide suggests against using `_` in operation and function names. In this case, I'd suggest that since this operation does more than just applying the identity operation, but also checks that the operation is correct, it may be helpful to rename to something like `CheckApplyUnitaryAppliesIdentity`. While that is a bit verbose, this operation isn't directly called in user code (not part of the Q# API), and verbose names can be helpful in reading through error logs.

fedimser replied:
Changed all test cases names to CheckApplyUnitaryApplies...

cgranade reviewed (changes requested):
Thanks for doing this, we really appreciate your contribution! I've marked as request changes for some of the required modifications (e.g.: copyright headers), but I'm happy to help expand on or implement any of my comments.

Since this PR modifies the Q# API, I'll also go on and schedule it for the next [API Review Meeting](https://github.com/microsoft/QuantumLibraries/projects/1), and will comment here accordingly following that meeting. Thank you!

msoeken commented on Standard/src/Synthesis/TwoLevelUnitary.cs:
```diff
@@ -0,0 +1,71 @@
+        }
+
+        // Converts to tuple to be passed to Q#.
+        public (IQArray<IQArray<Quantum.Math.Complex>>, long, long) toQsharp()
+        {
+            return (MatrixUtils.matrixToQs(this.mx), this.i1, this.i2);
+        }
```
@cgranade, should it be `=>` instead of `=` in line 66?

cgranade replied:
Yes, absolutely. Thanks for catching my typo, @msoeken!

cgranade replied:
I've edited the suggestion accordingly.

fedimser replied:
Done.

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.qs:
```diff
@@ -9,48 +9,40 @@ namespace Microsoft.Quantum.Synthesis {
         let lmbda = ArgComplex(u[0][0]);
         let mu = ArgComplex(u[0][1]);
-        if (AbsD(mu-lmbda) > 1e-9) { Rz(mu - lmbda, qubit); }
+        if (AbsD(mu - lmbda) > 1e-9) { Rz(mu - lmbda, qubit); }
```
Can we also use 1e-10 here as tolerance value?

fedimser replied:
Done.

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.qs:
```diff
@@ -9,48 +9,40 @@ namespace Microsoft.Quantum.Synthesis {
-        if (AbsD(mu-lmbda) > 1e-9) { Rz(mu - lmbda, qubit); }
+        if (AbsD(mu - lmbda) > 1e-9) { Rz(mu - lmbda, qubit); }
         if (AbsD(theta) > 1e-9) { Ry(-2.0 * theta, qubit); }
         if (AbsD(lmbda + mu) > 1e-9) { Rz(- lmbda - mu, qubit); }
```
```suggestion
        if (AbsD(lmbda + mu) > 1e-9) { Rz(-lmbda - mu, qubit); }
```

fedimser replied:
Done.

msoeken reviewed (commented):
Looks great to me, thanks! I left some few comments.

fedimser commented:
Hi Mathias,

I addressed your comments.
Also I have found a bug in my algorithm, which I fixed in my Python repository, as well as here.
I added one more test for 8x8 matrix which is equivalent to 4 controlled gates (this test helped me find a bug).

One more thing to double check is that my algorithm indeed applies matrix to little-endian register. The easiest way to check that is to see what is matrix for CNOT gate, and I have a test for that.
So, if you give this matrix to ApplyUnitary:
[[1, 0, 0, 0],
 [0, 0, 0, 1],
 [0, 0, 1, 0],
 [0, 1, 0, 0]],
then it will apply operation equivalent to CNOT(q[0], q[1]).
We can see that it swaps states 1 and 3, i.e. states with the lowest bit equal to 1. Which means qubit corresponding to thelowest bit is control qubit. And we see that control qubit is q[0]. So, the qubit with index 0 encodes the lowest bit, as it should.

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.cs:
```diff
@@ -48,28 +49,45 @@ private static List<TwoLevelUnitary> TwoLevelDecompose(Complex[,] A)
+                            // to multiplying by identity matrix.
+                            mx = new Complex[,] { { 1.0, 0.0 }, { 0.0, 1.0 } };
+                            // But if it's last in a row, ensure that diagonal element will be 1.
+                            if (j == i + 1) {
+                                mx = new Complex[,] { { 1/a, 0.0 }, { 0.0, a } };
+                            }                            
```
```suggestion
                            if (j == i + 1)
                            {
                                mx = new Complex[,] { { 1/a, 0.0 }, { 0.0, a } };
                            }                            
```

fedimser replied:
Done.

msoeken commented on Standard/src/Synthesis/UnitaryDecomposition.cs:
```diff
@@ -48,28 +49,45 @@ private static List<TwoLevelUnitary> TwoLevelDecompose(Complex[,] A)
+                            // Just swap columns with Pauli X matrix.
+                            mx = new Complex[,] { { 0.0, 1.0 }, { 1.0, 0.0 } };
+                            // But if it's last in a row, ensure that diagonal element will be 1.
+                            if (j == i + 1) {
+                                mx = new Complex[,] { { 0.0, b }, { 1/b, 0.0 } };
+                            } 
```
```suggestion
                            if (j == i + 1)
                            {
                                mx = new Complex[,] { { 0.0, b }, { 1/b, 0.0 } };
                            } 
```

fedimser replied:
Done.

msoeken commented on Standard/tests/Synthesis/UnitaryDecompositionTest.qs:
```diff
@@ -134,4 +132,30 @@ namespace Microsoft.Quantum.Tests {
+    }   
+
+    @Test("QuantumSimulator")
+    operation CheckApplyUnitaryAppliesFourControlledGates () : Unit {
```
```suggestion
    operation CheckApplyUnitaryAppliesFourControlledGates() : Unit {
```

fedimser replied:
Done.

msoeken reviewed (commented):
Just some formatting suggestions.

cgranade commented on Standard/src/Synthesis/UnitaryDecomposition.cs:
```diff
@@ -0,0 +1,172 @@
+
+            // Returns list of two-level unitary matrices, which multiply to A.
+            // Every matrix has indices differing in exactly 1 bit.
+            private static List<TwoLevelUnitary> twoLevelDecomposeGray(Complex[,] A)
+            {
+                int n = A.GetLength(0);
+                Debug.Assert(A.GetLength(1) == n, "Matrix is not square.");
+                Debug.Assert(MatrixUtils.IsMatrixUnitary(A), "Matrix is not unitary.");
+
+                int[] perm = GrayCode(n);
+                A = PermuteMatrix(A, perm);
+                List<TwoLevelUnitary> result = TwoLevelDecompose(A);
+                foreach (TwoLevelUnitary matrix in result)
+                {
+                    matrix.ApplyPermutation(perm);
+                    matrix.OrderIndices();
+                }
+                return result;
+            }
```
You may not need to collect in a `List` here, either. Using the `yield return` statement (very similar to Python's `yield`), you can write this out directly as an enumerable over `TwoLevelUnitary`:

```suggestion
            private static IEnumerable<TwoLevelUnitary> twoLevelDecomposeGray(Complex[,] A)
            {
                int n = A.GetLength(0);
                Debug.Assert(A.GetLength(1) == n, "Matrix is not square.");
                Debug.Assert(MatrixUtils.IsMatrixUnitary(A), "Matrix is not unitary.");

                int[] perm = GrayCode(n);
                A = PermuteMatrix(A, perm);
                foreach (TwoLevelUnitary matrix in  TwoLevelDecompose(A))
                {
                    matrix.ApplyPermutation(perm);
                    matrix.OrderIndices();
                    yield return matrix;
                }
            }
```

Using the same technique in `TwoLevelDecompose` should allow you to avoid allocating a list for any intermediate steps, only allocating in the constructor call at line 168.

fedimser replied:
Done.

cgranade reviewed (approved):
This looks really good, thanks for the contribution! I've left a few last comments as suggestions that would be good to bring in before merging this to `main`, but otherwise I think this is getting pretty close now that the API review has completed.

Thank you again!

fedimser commented on Standard/src/Synthesis/UnitaryDecomposition.cs:
```diff
@@ -0,0 +1,172 @@
+
+            // Returns list of two-level unitary matrices, which multiply to A.
+            // Every matrix has indices differing in exactly 1 bit.
+            private static List<TwoLevelUnitary> twoLevelDecomposeGray(Complex[,] A)
```
Done.

fedimser commented on Standard/src/Synthesis/UnitaryDecomposition.qs:
```diff
@@ -0,0 +1,78 @@
+    /// ## qubits
+    /// Qubits to which apply the operation - register of length n.
+    operation ApplyUnitary(unitary: Complex[][], qubits : LittleEndian) : Unit is Adj + Ctl {
+        if (Length(unitary) != 1 <<< Length(qubits!)) {
+            fail "Matrix size is not consistent with register length.";
+        }
```
Done.

fedimser commented on Standard/src/Synthesis/UnitaryDecomposition.qs:
```diff
@@ -0,0 +1,78 @@
+    /// If matrix is not unitary or not of suitable size, throws an exception.
+    /// ## qubits
+    /// Qubits to which apply the operation - register of length n.
+    operation ApplyUnitary(unitary: Complex[][], qubits : LittleEndian) : Unit is Adj + Ctl {
```
Done. 
Added internal function SquareMatrixFact.

Strictly speaking, I would also have to add a check that matrix is unitary (which is already present in Q#). But it's non-trivial  and it's complexity is O(N^3), so I would like avoid implementing and running it twice. 
So I hope it's fine to check that matrix is unitary only in C#.

fedimser commented on Standard/src/Synthesis/UnitaryDecomposition.cs:
```diff
@@ -0,0 +1,172 @@
+            private static List<TwoLevelUnitary> twoLevelDecomposeGray(Complex[,] A)
+            {
+                int n = A.GetLength(0);
+                Debug.Assert(A.GetLength(1) == n, "Matrix is not square.");
```
Added check that matrix is square in Q#.

fedimser commented:
@cgranade, I addressed your comments. I think this PR is now ready to be merged.

cgranade reviewed (approved):
Looks good, thank you for the contribution! ♥

cgranade commented:
/azp run

cgranade commented:
/azp run
